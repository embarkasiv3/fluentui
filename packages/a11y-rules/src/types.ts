// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type PropValue = string | number | boolean | any;
export type Props = { [name: string]: PropValue };
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AccessibilityBehavior = (props: Props) => any;

export interface Rule {
  stringify: () => string;
}

export interface Slot {
  name: string;
  props: Props[];
  expectAttribute: boolean;
  expectedAttribute: string;
  expectedValue: PropValue;
  description: string;
  hidden: boolean;
  afterEventData: Event;
  checkClick: boolean;
  checkSpaceKeyPressed: boolean;
  checkEnterKeyPressed: boolean;
  checkOnClickWasExecuted: boolean;
}
