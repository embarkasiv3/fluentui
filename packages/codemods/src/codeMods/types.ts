import { SourceFile } from 'ts-morph';

export interface CodeModResult {
  success?: boolean;
}

export interface Codemod<T = SourceFile> {
  /**
   * Each type of codemod can have multiple versions which work on different versions of its targeted package.
   * Must be valid semver.
   * TODO: Currently version is not being checked. Implement this fully or remove.
   */
  version?: string;
  /**
   * A string to help identify the codemod.
   */
  name: string;
  /**
   * The actual function that should be run on any given file.
   * TODO, is there a possibility of codemods that would need to execute over mutiple files?
   */
  run: (file: T) => CodeModResult;
  /**
   * If not enabled, then this mod will not be conisdered to run. Only enable it once it's ready
   * to be applied in real world scenarios
   */
  enabled?: boolean;
}
