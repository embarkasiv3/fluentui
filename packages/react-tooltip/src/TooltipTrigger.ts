export * from './components/TooltipTrigger/index';
