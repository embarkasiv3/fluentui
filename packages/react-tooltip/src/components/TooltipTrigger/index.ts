export * from './TooltipTrigger';
export * from './TooltipTrigger.types';
export * from './renderTooltipTrigger';
export * from './useTooltipTrigger';
