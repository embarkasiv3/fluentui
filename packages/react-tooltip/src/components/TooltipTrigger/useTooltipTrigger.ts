import * as React from 'react';
import { makeMergeProps, resolveShorthandProps, useId } from '@fluentui/react-utilities';
import { TooltipTriggerChildProps, TooltipTriggerProps, TooltipTriggerState } from './TooltipTrigger.types';
import { TooltipImperativeHandle } from '../Tooltip/Tooltip.types';
import { useTooltipContext } from '../TooltipProvider/useTooltipContext';

/**
 * Names of the shorthand properties in TooltipTriggerProps
 * {@docCategory TooltipTrigger}
 */
export const tooltipTriggerShorthandProps = ['tooltip'] as const;

const mergeProps = makeMergeProps<TooltipTriggerState>({ deepMerge: tooltipTriggerShorthandProps });

/**
 * Create the state required to render TooltipTrigger.
 *
 * The returned state can be modified with hooks such as useTooltipTriggerStyles,
 * before being passed to renderTooltipTrigger.
 *
 * @param props - props from this instance of TooltipTrigger
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory TooltipTrigger}
 */
export const useTooltipTrigger = (
  props: TooltipTriggerProps,
  defaultProps?: TooltipTriggerProps,
): TooltipTriggerState => {
  const tooltipRef = React.useRef<TooltipImperativeHandle | null>(null);

  const { TooltipComponent, tooltipManager } = useTooltipContext();

  const state = mergeProps(
    {
      children: props.children,
      tooltip: {
        as: TooltipComponent,
        id: useId('tooltip-'),
        componentRef: tooltipRef,
      },
      tooltipManager,
      tooltipRef,
    },
    defaultProps && resolveShorthandProps(defaultProps, tooltipTriggerShorthandProps),
    resolveShorthandProps(props, tooltipTriggerShorthandProps),
  );

  mergeProps(state, {
    tooltip: {
      // Some TooltipProps are on TooltipTrigger for convenience, and need to be added to the tooltip
      position: state.position,
      align: state.align,
      subtle: state.subtle,
      noArrow: state.noArrow,
      offset: state.offset,
      onPointerEnter: mergeEventCallbacks(() => tooltipManager?.onPointerEnterTooltip(), state.tooltip.onPointerEnter),
      onPointerLeave: mergeEventCallbacks(() => tooltipManager?.onPointerLeaveTooltip(), state.tooltip.onPointerLeave),
    },
  });

  if (typeof state.children === 'function') {
    // If a render function was passed in as the child, pass the props to the function
    state.children = state.children(extraChildProps(state)) as TooltipTriggerState['children'];
  } else {
    const child = React.Children.only(state.children);
    if (React.isValidElement(child) && child.type !== React.Fragment) {
      // Attach the extra props by cloning the child
      state.children = React.cloneElement(child, extraChildProps(state, child.props));
    } else if (process.env.NODE_ENV !== 'production') {
      throw new Error(
        'TooltipTrigger has unsupported children. It can only contain a single React element, or a render function.',
      );
    }
  }

  return state;
};

/**
 * Create the extra props that are appended to the tooltip
 */
const extraChildProps = (
  state: TooltipTriggerState,
  childProps?: React.HTMLAttributes<HTMLElement>,
): TooltipTriggerChildProps => {
  const showTooltipHandler = (ev: React.SyntheticEvent<HTMLElement>) => {
    if (state.tooltipRef.current) {
      const target = state.targetRef?.current ?? ev.currentTarget;
      if (state.onlyIfTruncated) {
        const truncated = target.clientWidth < target.scrollWidth || target.clientHeight < target.scrollHeight;
        if (!truncated) {
          return;
        }
      }

      state.tooltipManager?.showTooltip({
        tooltip: state.tooltipRef.current,
        trigger: ev.currentTarget,
        target,
        showDelay: state.showDelay,
        hideDelay: state.hideDelay,
      });
    }
  };

  const hideTooltipHandler = (ev: React.SyntheticEvent<HTMLElement>) => {
    state.tooltipManager?.hideTooltip(ev.currentTarget);
  };

  return {
    onPointerEnter: mergeEventCallbacks(childProps?.onPointerEnter, showTooltipHandler),
    onPointerLeave: mergeEventCallbacks(childProps?.onPointerLeave, hideTooltipHandler),
    onFocus: mergeEventCallbacks(childProps?.onFocus, showTooltipHandler),
    onBlur: mergeEventCallbacks(childProps?.onBlur, hideTooltipHandler),

    // If the tooltip is a label, it sets aria-labelledby to the tooltip's ID instead of aria-describedby
    [state.type === 'label' ? 'aria-labelledby' : 'aria-describedby']: state.tooltip.id,
  };
};

const mergeEventCallbacks = <Event>(...callbacks: (((ev: Event) => void) | undefined)[]) => {
  return (ev: Event) => callbacks.forEach(callback => callback?.(ev));
};
