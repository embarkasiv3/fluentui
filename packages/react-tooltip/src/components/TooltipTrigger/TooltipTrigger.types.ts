import * as React from 'react';
import { RequiredProps, ResolvedShorthandProps, ShorthandProps } from '@fluentui/react-utilities';
import { tooltipTriggerShorthandProps } from './useTooltipTrigger';
import { TooltipImperativeHandle, TooltipProps } from '../Tooltip/Tooltip.types';
import { TooltipManager } from '../TooltipProvider/useTooltipManager';

/**
 * {@docCategory TooltipTrigger}
 */
export interface TooltipTriggerProps
  extends Pick<TooltipProps, 'position' | 'align' | 'subtle' | 'noArrow' | 'offset'> {
  /**
   * The child of TooltipTrigger is the element that triggers the tooltip. It will
   * have additional properties added, including events and aria properties.
   * Alternatively, children can be a render function that takes the props and adds
   * them to the appropriate elements.
   */
  children:
    | React.ReactElement<React.HTMLAttributes<HTMLElement>>
    | ((props: TooltipTriggerChildProps) => React.ReactNode);

  /**
   * The content of the tooltip.
   */
  tooltip: ShorthandProps<TooltipProps>;

  /**
   * Determines whether the tooltip is being used as the trigger's label or description.
   * This determines whether to set aria-describedby or aria-labelledby on the trigger element.
   *
   * @defaultvalue description
   */
  type?: 'description' | 'label';

  /**
   * Delay before the tooltip is shown, in milliseconds
   *
   * @defaultvalue 250
   */
  showDelay?: number;

  /**
   * Delay before the tooltip is hidden, in milliseconds
   *
   * @defaultvalue 250
   */
  hideDelay?: number;

  /**
   * Only show the tooltip if the target element's children are truncated (overflowing).
   */
  onlyIfTruncated?: boolean;

  /**
   * A ref to an element that the tooltip should be anchored to.
   *
   * If not specified, the tooltip will point to the same element that triggered it, which is the common use case.
   */
  targetRef?: React.RefObject<HTMLElement>;
}

/**
 * The props that are added to the child of the TooltipTrigger
 */
export type TooltipTriggerChildProps = Pick<
  React.HTMLAttributes<HTMLElement>,
  'onPointerEnter' | 'onPointerLeave' | 'onFocus' | 'onBlur' | 'aria-describedby' | 'aria-labelledby'
>;

/**
 * Names of the shorthand properties in TooltipTriggerProps
 * {@docCategory TooltipTrigger}
 */
export type TooltipTriggerShorthandProps = typeof tooltipTriggerShorthandProps[number];

/**
 * Names of TooltipTriggerProps that have a default value in useTooltipTrigger
 * {@docCategory TooltipTrigger}
 */
export type TooltipTriggerDefaultedProps = 'tooltip';

/**
 * {@docCategory TooltipTrigger}
 */
export type TooltipTriggerState = RequiredProps<
  ResolvedShorthandProps<
    TooltipTriggerProps & {
      tooltipManager: TooltipManager | undefined;
      tooltipContainer: HTMLElement | undefined;
      tooltipRef: React.MutableRefObject<TooltipImperativeHandle | null>;
    },
    TooltipTriggerShorthandProps
  >,
  TooltipTriggerDefaultedProps
>;
