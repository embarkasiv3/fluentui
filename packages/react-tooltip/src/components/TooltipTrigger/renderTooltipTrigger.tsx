import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { getSlots, isSSR } from '@fluentui/react-utilities';
import { TooltipTriggerState } from './TooltipTrigger.types';
import { tooltipTriggerShorthandProps } from './useTooltipTrigger';

/**
 * Render the final JSX of TooltipTrigger
 * {@docCategory TooltipTrigger}
 */
export const renderTooltipTrigger = (state: TooltipTriggerState) => {
  const { slots, slotProps } = getSlots(state, tooltipTriggerShorthandProps);

  return (
    <>
      {state.children}
      {isSSR() || state.tooltipContainer === undefined
        ? null
        : ReactDOM.createPortal(<slots.tooltip {...slotProps.tooltip} />, state.tooltipContainer)}
    </>
  );
};
