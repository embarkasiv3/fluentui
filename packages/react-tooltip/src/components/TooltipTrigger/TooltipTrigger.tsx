import * as React from 'react';
import { useTooltipTrigger } from './useTooltipTrigger';
import { TooltipTriggerProps } from './TooltipTrigger.types';
import { renderTooltipTrigger } from './renderTooltipTrigger';

/**
 * Wrap an element that will trigger a tooltip when focused or moused over.
 *
 * {@docCategory TooltipTrigger}
 */
export const TooltipTrigger: React.FunctionComponent<TooltipTriggerProps> = props => {
  const state = useTooltipTrigger(props);
  return renderTooltipTrigger(state);
};

TooltipTrigger.displayName = 'TooltipTrigger';
