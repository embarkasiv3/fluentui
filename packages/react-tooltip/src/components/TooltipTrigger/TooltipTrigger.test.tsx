import * as React from 'react';
import { TooltipTrigger } from './TooltipTrigger';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { TooltipTriggerProps } from './TooltipTrigger.types';

describe('TooltipTrigger', () => {
  isConformant<TooltipTriggerProps>({
    Component: TooltipTrigger,
    displayName: 'TooltipTrigger',
    requiredProps: { children: <div /> },
    disabledTests: ['component-handles-ref', 'component-has-root-ref', 'component-handles-classname'],
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for TooltipTrigger in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(
      <TooltipTrigger tooltip="Hello world">
        <button>hello world</button>
      </TooltipTrigger>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
