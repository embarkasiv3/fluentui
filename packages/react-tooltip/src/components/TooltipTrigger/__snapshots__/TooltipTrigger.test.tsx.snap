// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipTrigger renders a default state 1`] = `
<button
  aria-describedby="tooltip-1"
  onBlur={[Function]}
  onFocus={[Function]}
  onPointerEnter={[Function]}
  onPointerLeave={[Function]}
>
  hello world
</button>
`;
